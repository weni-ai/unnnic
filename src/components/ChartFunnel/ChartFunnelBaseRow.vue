<template>
  <svg
    class="unnnic-chart-funnel-base-rows"
    :viewBox="viewBox"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      v-for="(row, index) of rows"
      :key="row.title + index"
      class="unnnic-chart-funnel-base-rows__row"
    >
      <path
        class="row__path"
        :d="row.pathD"
      />
      <text
        x="50%"
        :y="row.titleY"
        class="row__title"
      >
        {{ row.title }}
      </text>
      <text
        x="50%"
        :y="row.descriptionY"
        class="row__description"
      >
        {{ row.description }}
      </text>
    </g>
  </svg>
</template>

<script>
export default {
  name: 'UnnnicChartFunnelThreeRows',

  props: {
    rows: {
      type: Array,
      required: true,
    },

    viewBox: String,
  },
};
</script>

<style lang="scss" scoped>
@import '../../assets/scss/unnnic.scss';

.unnnic-chart-funnel-base-rows {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;

  &__row {
    .row__title,
    .row__description {
      fill: $unnnic-color-neutral-white;
      font-family: $unnnic-font-family-secondary;
      text-anchor: middle;
    }

    .row__title {
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-bold;
    }

    .row__description {
      font-size: $unnnic-font-size-body-md;
    }
  }
}
</style>
