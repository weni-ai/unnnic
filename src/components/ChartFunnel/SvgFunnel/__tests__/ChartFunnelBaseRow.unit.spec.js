import { mount } from '@vue/test-utils';
import { describe, expect, test, beforeEach } from 'vitest';
import ChartFunnelBaseRow from '@/components/ChartFunnel/SvgFunnel/ChartFunnelBaseRow.vue';

describe('ChartFunnelBaseRow.vue', () => {
  let wrapper;
  const mockRows = [
    {
      title: '100%',
      description: 'First Level',
      pathD: 'M0 0 L100 0 L80 50 L20 50 Z',
      titleY: '20',
      descriptionY: '35',
    },
    {
      title: '75%',
      description: 'Second Level',
      pathD: 'M20 50 L80 50 L60 100 L40 100 Z',
      titleY: '70',
      descriptionY: '85',
    },
    {
      title: '50%',
      description: 'Third Level',
      pathD: 'M40 100 L60 100 L50 150 L50 150 Z',
      titleY: '120',
      descriptionY: '135',
    },
  ];

  beforeEach(() => {
    wrapper = mount(ChartFunnelBaseRow, {
      props: {
        rows: mockRows,
        viewBox: '0 0 100 150',
      },
    });
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('validates required props', () => {
    const { rows } = wrapper.vm.$options.props;
    expect(rows.required).toBe(true);
    expect(rows.type).toBe(Array);
  });

  test('validates optional viewBox prop', () => {
    const { viewBox } = wrapper.vm.$options.props;
    expect(viewBox.type).toBe(String);
    expect(viewBox.required).toBeTruthy();
  });

  test('applies correct viewBox to svg element', () => {
    const svg = wrapper.find('svg');
    expect(svg.attributes('viewBox')).toBe('0 0 100 150');
  });

  test('renders correct number of rows', () => {
    const rows = wrapper.findAll('.unnnic-chart-funnel-base-rows__row');
    expect(rows).toHaveLength(mockRows.length);
  });

  describe('row elements rendering', () => {
    test('renders paths with correct d attribute', () => {
      const paths = wrapper.findAll('.row__path');
      paths.forEach((path, index) => {
        expect(path.attributes('d')).toBe(mockRows[index].pathD);
      });
    });

    test('renders titles with correct text and position', () => {
      const titles = wrapper.findAll('.row__title');
      titles.forEach((title, index) => {
        expect(title.text()).toBe(mockRows[index].title);
        expect(title.attributes('x')).toBe('50%');
        expect(title.attributes('y')).toBe(mockRows[index].titleY);
      });
    });

    test('renders descriptions with correct text and position', () => {
      const descriptions = wrapper.findAll('.row__description');
      descriptions.forEach((description, index) => {
        expect(description.text()).toBe(mockRows[index].description);
        expect(description.attributes('x')).toBe('50%');
        expect(description.attributes('y')).toBe(mockRows[index].descriptionY);
      });
    });
  });

  test('handles empty rows array', async () => {
    await wrapper.setProps({ rows: [] });
    const rows = wrapper.findAll('.unnnic-chart-funnel-base-rows__row');
    expect(rows).toHaveLength(0);
  });

  test('handles missing viewBox prop', async () => {
    const wrapperWithoutViewBox = mount(ChartFunnelBaseRow, {
      props: {
        rows: mockRows,
      },
    });
    const svg = wrapperWithoutViewBox.find('svg');
    expect(svg.exists()).toBe(true);
    expect(svg.attributes('viewbox')).toBeFalsy();
  });

  test('handles rows with missing properties', async () => {
    const incompleteRows = [
      {
        title: '100%',
        pathD: 'M0 0 L100 0 L80 50 L20 50 Z',
        titleY: '20',
      },
    ];

    await wrapper.setProps({ rows: incompleteRows });
    const row = wrapper.find('.unnnic-chart-funnel-base-rows__row');
    const title = row.find('.row__title');
    const description = row.find('.row__description');
    const path = row.find('.row__path');

    expect(title.text()).toBe('100%');
    expect(description.text()).toBe('');
    expect(path.attributes('d')).toBe(incompleteRows[0].pathD);
  });

  test('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('verifies SVG namespace', () => {
    const svg = wrapper.find('svg');
    expect(svg.attributes('xmlns')).toBe('http://www.w3.org/2000/svg');
    expect(svg.attributes('fill')).toBe('none');
  });

  test('maintains component class name', () => {
    const svg = wrapper.find('svg');
    expect(svg.classes()).toContain('unnnic-chart-funnel-base-rows');
  });

  test('preserves row order', () => {
    const rows = wrapper.findAll('.unnnic-chart-funnel-base-rows__row');
    rows.forEach((row, index) => {
      const title = row.find('.row__title');
      expect(title.text()).toBe(mockRows[index].title);
    });
  });

  describe('edge cases', () => {
    test('handles rows with special characters in text', async () => {
      const rowsWithSpecialChars = [
        {
          title: '100% & <special>',
          description: 'Test & <special>',
          pathD: 'M0 0 L100 0 L80 50 L20 50 Z',
          titleY: '20',
          descriptionY: '35',
        },
      ];

      await wrapper.setProps({ rows: rowsWithSpecialChars });
      const title = wrapper.find('.row__title');
      const description = wrapper.find('.row__description');

      expect(title.text()).toBe('100% & <special>');
      expect(description.text()).toBe('Test & <special>');
    });

    test('handles very long text content', async () => {
      const rowsWithLongText = [
        {
          title: 'A'.repeat(50),
          description: 'B'.repeat(100),
          pathD: 'M0 0 L100 0 L80 50 L20 50 Z',
          titleY: '20',
          descriptionY: '35',
        },
      ];

      await wrapper.setProps({ rows: rowsWithLongText });
      const title = wrapper.find('.row__title');
      const description = wrapper.find('.row__description');

      expect(title.text()).toBe('A'.repeat(50));
      expect(description.text()).toBe('B'.repeat(100));
    });
  });
});
