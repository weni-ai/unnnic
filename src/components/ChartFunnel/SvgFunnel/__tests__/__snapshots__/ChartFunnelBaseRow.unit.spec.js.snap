// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChartFunnelBaseRow.vue > matches snapshot 1`] = `
"<svg data-v-38919cea="" class="unnnic-chart-funnel-base-rows" viewBox="0 0 100 150" fill="none" xmlns="http://www.w3.org/2000/svg">
  <g data-v-38919cea="" class="unnnic-chart-funnel-base-rows__row">
    <path data-v-38919cea="" class="row__path" d="M0 0 L100 0 L80 50 L20 50 Z"></path><text data-v-38919cea="" x="50%" y="20" class="row__title">100%</text><text data-v-38919cea="" x="50%" y="35" class="row__description">First Level</text>
  </g>
  <g data-v-38919cea="" class="unnnic-chart-funnel-base-rows__row">
    <path data-v-38919cea="" class="row__path" d="M20 50 L80 50 L60 100 L40 100 Z"></path><text data-v-38919cea="" x="50%" y="70" class="row__title">75%</text><text data-v-38919cea="" x="50%" y="85" class="row__description">Second Level</text>
  </g>
  <g data-v-38919cea="" class="unnnic-chart-funnel-base-rows__row">
    <path data-v-38919cea="" class="row__path" d="M40 100 L60 100 L50 150 L50 150 Z"></path><text data-v-38919cea="" x="50%" y="120" class="row__title">50%</text><text data-v-38919cea="" x="50%" y="135" class="row__description">Third Level</text>
  </g>
</svg>"
`;
